File: backend/api/app/domain/use_cases/chat/create.py
from core.use_cases.use_case import UseCase
from app.domain.repositories.chat import BaseRepository
from core.common.equatable import Equatable
from core.common.either import Either
from core.errors.failure import Failure
from app.domain.entities.chat import ChatEntity
from app.domain.entities.message import Message

class Params(Equatable):
    def __init__(self, message: Message) -> None:
        self.message = message


class CreateChat(UseCase[ChatEntity]):
    def __init__(self, repository: BaseRepository):
        self.repository = repository
    
    async def __call__(self, params: Params) -> Either[Failure, ChatEntity]:
        return await self.repository.create_chat(params.message)

File: mobile/lib/features/architect/domains/use_cases/chat/create.dart
import 'package:dartz/dartz.dart';
import 'package:equatable/equatable.dart';

import '../../../../../core/errors/failure.dart';
import '../../../../../core/use_cases/usecase.dart';
import '../../entities/chat.dart';
import '../../repositories/chat.dart';

class CreateChat extends UseCase<Chat, Params> {
  CreateChat(
    this.repository,
  );

  final ChatRepository repository;

  @override
  Future<Either<Failure, Chat>> call(Params params) async {
    return await repository.create(
      payload: params.payload,
      userId: params.userId,
      model: params.model,
    );
  }
}

class Params extends Equatable {
  const Params({
    required this.payload,
    required this.userId,
    required this.model,
  });

  final Map<String, dynamic> payload;
  final String userId;
  final String model;

  @override
  List<Object?> get props => [
        payload,
        userId,
        model,
      ];
}
